### PR TITLE
feat(amplify-cli-core): Allow cloudformation functions (e.g. Fn::Sub, Fn::Join) in custom-policies.json resources

### PR DESCRIPTION
**Pull request by juniarz**
#### Description of changes

Allow objects in custom-policies.json

E.g.
```json
[
  {
    "Effect": "Allow",
    "Action": [
      "ses:SendTemplatedEmail"
    ],
    "Resource": [
      {
        "Fn::Sub": [
          "arn:aws:ses:${region}:${account}:identity/*",
          {
            "region": "us-east-1",
            "account": {
              "Ref": "AWS::AccountId"
            }
          }
        ]
      }
    ]
  }
]
```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mattfysh** [comment]
Will this also allow us to reference parameters from the sibling template.json file?

**Comment by mattfysh** [comment]
This seems to be working in the latest version of Amplify CLI... able to use Fn::Join and also reference parameters from template.json

**Comment by juniarz** [comment]
> Will this also allow us to reference parameters from the sibling template.json file?

Hello, yes this PR allows us to use cloudformation functions in custom-policies, as such reference should also work.

**Comment by juniarz** [comment]
Hi, the PR have been pending review for the last 2 months... any updates?

It is annoying having to hardcode ARNs in the custom-policies.json.

It does not work with multi environments...

File: packages/amplify-cli-core/src/customPoliciesUtils.ts [hunk 1]
```diff
@@ -124,6 +132,10 @@ function validateCustomPolicies(data: CustomIAMPolicies, categoryName: string, r
     let errorMessage = '';
 
     for (const resource of resources) {
+      if (typeof resource === 'object') {
```
Comment (jhockett) [inline packages/amplify-cli-core/src/customPoliciesUtils.ts:135]:
```suggestion
      if (typeof resource !== 'string') {
```

File: packages/amplify-cli-core/src/customPoliciesUtils.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ export type CustomIAMPolicies = CustomIAMPolicy[];
 export type CustomIAMPolicy = {
   Action: string[];
   Effect?: string;
-  Resource: string[];
+  Resource: (string | object)[];
```
Comment (jhockett) [inline packages/amplify-cli-core/src/customPoliciesUtils.ts:14]: Can you change this to `Resource: (string | $TSObject)[];`?

You will need to import the `$TSObject` type from `amplify-cli-core/src/index.ts`

**Comment by jhockett** [comment]
Sorry for the delayed review @juniarz and thanks so much for the contribution! It essentially LGTM. I left a couple of comments to remove the use of `object` as a type.

**Review by jhockett — CHANGES_REQUESTED** [review]

File: packages/amplify-cli-core/src/customPoliciesUtils.ts (hunk 2 above)
Comment (juniarz) [reply packages/amplify-cli-core/src/customPoliciesUtils.ts]: Implemented in latest commit https://github.com/aws-amplify/amplify-cli/pull/9076/commits/343b339ebfb58c41d01592a0aa25d1508370cd0a

**Review by jhockett — APPROVED** [review]
LGTM!

**Review by ammarkarachi — APPROVED** [review]

**Comment by juniarz** [comment]
Thank you john!

**Pull request merged**